packages/jest-haste-map/src/index.js

```diff
@@ -681,6 +706,49 @@ class HasteMap extends EventEmitter {
     });
   }
 
+  /**
+   * This function should be called when the file under `filePath` is removed
+   * or changed. When that happens, we want to figure out if that file was
+   * part of a group of files that had the same ID. If it was, we want to
+   * remove it from the group. Furthermore, if there is only one file
+   * remaining in the group, then we want to restore that single file as the
+   * correct resolution for its ID, and cleanup the duplicates index.
+   */
+  _recoverDuplicates(
+    hasteMap: InternalHasteMap,
+    filePath: string,
+    moduleName: string,
+  ) {
+    let dupsByPl = hasteMap.duplicates[moduleName];
+    if (dupsByPl == null) {
+      return;
+    }
+    const platform =
+      getPlatformExtension(filePath, this._options.platforms) ||
+      H.GENERIC_PLATFORM;
+    let dups = dupsByPl[platform];
+    if (dups == null) {
+      return;
+    }
+    dupsByPl = hasteMap.duplicates[moduleName] = (copy(dupsByPl): any);
+    dups = dupsByPl[platform] = (copy(dups): any);
+    const dedupType = dups[filePath];
+    delete dups[filePath];
```

Comment:
Alright, the remaining thing I'm confused about is why the currently changed file gets removed from the duplicate map on this line.